### PR TITLE
[Front-End] User Types and Privileges Setup

### PR DESCRIPTION
**Pull request by nex-pixel**
## Feature Description/Implementation
Added a new page called UserAppLayer for determining which type of priviledge the user has.
Added a new folder with permissions functions and code to determine whether a particular action is allowed or not
Refactored Navbar and Router as well as index.tsx to align with user changes
Refactored login page to allow for passing of user types up to user layer
Refactored User detail view for Success messages and restrictions on modifying self or "admin"
Added restrictions to components that according to evolution 3 need to be hidden. (may have missed a couple)

## Everything Else
Solution for refresh token was found, but has yet to be implemented.

### CODE REVIEW

File: frontend/src/components/navigation/Router.tsx [hunk 1]
```diff
@@ -18,39 +17,46 @@ import PurchaseOrderList from "../../pages/purchases/POList";
...
 
-const WithNavBar = () => {
-  return (
-    <>
-      <NavigationBar />
-      <Outlet />
-    </>
-  );
-};
+interface NavbarProps {
```
Comment (caseyszilagyi) [inline frontend/src/components/navigation/Router.tsx:26]: We have NavbarProps here, and then NavigationBarProps in the Navbar file. They seem to have the exact same information too. I'm wondering if there's any way to get rid of one or the other? 

Maybe we can just get rid of this "WithNavBar" in the Router file, and just directly instantiate the <NavigationBar> in the Router. Then, we can just return an <Outlet> along with the NavBar in the Navbar file. Something similar to that, but I think we have an extra thing here that we don't really need

**Review by caseyszilagyi — CHANGES_REQUESTED** [review]
Looks good, a couple small things but overall I like the structure. I was wondering if this PR accounts for all special privileges that the "admin" account has, or if that will be needed later?

File: frontend/src/components/navigation/Navbar.tsx [hunk 2]
```diff
@@ -59,8 +66,9 @@ function NavigationBar() {
           label: "Log Out",
           icon: "pi pi-fw pi-sign-out",
           command: () => {
+            props.onLogout(undefined);
             localStorage.removeItem("accessToken");
-            navigate("/");
+            localStorage.removeItem("userID");
```
[the remark this answers is not in the document]
Comment (nex-pixel) [reply frontend/src/components/navigation/Navbar.tsx:71]: Login is not a direct path, its its own standalone page, that depending on the state of the current user, and access token status displayes the login page, or gives access to the router page. Its basically doing conditional logic between login page, and the rest of the website by checking user status with tokens and the current User variable in App.tsx

File: frontend/src/pages/vendors/VendorDetail.tsx [hunk 3]
```diff
@@ -186,14 +187,16 @@ export default function VendorDetail() {
...
-              {!isModifiable && (
-                <Button
-                  type="button"
-                  label="Edit"
-                  icon="pi pi-pencil"
-                  onClick={() => setIsModifiable(!isModifiable)}
-                />
-              )}
+              <Restricted to={"modify"}>
+                {!isModifiable && (
```
[the remark this answers is not in the document]
Comment (nex-pixel) [reply frontend/src/pages/vendors/VendorDetail.tsx:191]: Some components don't have a visible property for some reason. I think ToggleButton is one example, but yea I will try to go for that.

File: frontend/src/pages/users/UserDetail.tsx [hunk 4]
```diff
@@ -38,12 +38,20 @@ export default function VendorAdd() {
...
-        setIsAdmin(response.is_staff);
-      })
-      .catch(() => showFailure(toast, "Could not fetch user data"));
+    if (!isUserAddPage) {
+      USER_API.getUserDetail({ id: id! })
+        .then((response) => {
+          if (
+            id === localStorage.getItem("userID") ||
+            response.username === "admin"
+          ) {
```
[the remark this answers is not in the document]
Comment (nex-pixel) [reply frontend/src/pages/users/UserDetail.tsx:47]: Ok I extracted the conditional to be more informative of the function

File: frontend/src/App.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import { USER_API } from "./apis/users/UserAPI";
+import PermissionProvider from "./permissions/PermissionProvider";
+import { AccessType, administrator, user } from "./util/UserTypes";
+import LoginPage from "./pages/auth/LoginPage";
+
+function App() {
+  const [currentUser, setCurrentUser] = useState<AccessType | undefined>();
+
+  if (!currentUser) {
+    if (localStorage.getItem("accessToken") === null) {
```
[the remark this answers is not in the document]
Comment (nex-pixel) [reply frontend/src/App.tsx:12]: This is a temporary fix. The problem is right now that on page reload, all Javascript elements are re-executed. This means that currentUser is set to undefined, even though there is a accessToken. I just wanted to ensure that current functionality of many of the cancel button's still work. Later on, we might also need to check for refresh tokens here to renew the token if possible. If none of these is possible, go to login page. But yes, it works how you said it is right now.

File: frontend/src/components/navigation/Router.tsx (hunk 1 above)
Comment (nex-pixel) [reply frontend/src/components/navigation/Router.tsx:26]: Did it. Removed the unnecessary in between function.

File: frontend/src/components/navigation/Navbar.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (caseyszilagyi) [reply frontend/src/components/navigation/Navbar.tsx]: Makes sense

**Review by caseyszilagyi — APPROVED** [review]

**Pull request merged**